lib/SILGen/SILGenPoly.cpp

```diff
@@ -2015,29 +2015,49 @@ ResultPlanner::planTupleIntoDirectResult(AbstractionPattern innerOrigType,
 
   CanTupleType outerSubstTupleType = dyn_cast<TupleType>(outerSubstType);
 
-  // If the outer type is not a tuple, it must be optional.
+  // If the outer type is not a tuple, it must be optional or we are under
+  // opaque value mode
   if (!outerSubstTupleType) {
     CanType outerSubstObjectType =
       outerSubstType.getAnyOptionalObjectType();
-    assert(outerSubstObjectType &&
-           "inner type was a tuple but outer type was neither a tuple nor "
-           "optional");
-
-    auto someDecl = Gen.getASTContext().getOptionalSomeDecl();
-    SILType outerObjectType =
-        Gen.getSILType(outerResult).getAnyOptionalObjectType();
-    SILResultInfo outerObjectResult(outerObjectType.getSwiftRValueType(),
-                                    outerResult.getConvention());
-
-    // Plan to leave the tuple elements as a single direct outer result.
-    planTupleIntoDirectResult(innerOrigType, innerSubstType,
-                              outerOrigType.getAnyOptionalObjectType(),
-                              outerSubstObjectType,
-                              planData, outerObjectResult);
-
-    // Take that result and inject it into an optional.
-    addInjectOptionalDirect(someDecl, outerResult);
-    return;
+
+    if (outerSubstObjectType) {
+      auto someDecl = Gen.getASTContext().getOptionalSomeDecl();
+      SILType outerObjectType =
+          Gen.getSILType(outerResult).getAnyOptionalObjectType();
+      SILResultInfo outerObjectResult(outerObjectType.getSwiftRValueType(),
+                                      outerResult.getConvention());
+
+      // Plan to leave the tuple elements as a single direct outer result.
+      planTupleIntoDirectResult(innerOrigType, innerSubstType,
+                                outerOrigType.getAnyOptionalObjectType(),
+                                outerSubstObjectType, planData,
+                                outerObjectResult);
+
+      // Take that result and inject it into an optional.
+      addInjectOptionalDirect(someDecl, outerResult);
+      return;
+    } else {
+      assert(!Gen.silConv.useLoweredAddresses() &&
+             "inner type was a tuple but outer type was neither a tuple nor "
+             "optional nor are we under opaque value mode");
+      assert(outerSubstType->isAny());
+      auto &C = Gen.SGM.getASTContext();
+      auto layout = SILLayout::get(C, nullptr, SILField(outerSubstType, true));
+      auto canSILBoxTy = SILBoxType::get(C, layout, {});
+
+      AllocBoxInst *allocBox = Gen.B.createAllocBox(Loc, canSILBoxTy);
```

Comment:
Address lowering won't do anything with boxes. The Box2Stack pass will handle it. Ideally we won't be creating a box here. I just assumed it was more natural for SILGen to currently emit the box. If the thunk generator knows it's just for a return value, I'm not sure why it would be needed.